File: ProyectoCivitasRuby/lib/modelo/civitas_juego.rb
#encoding: UTF-8
require_relative 'dado'
require_relative 'estados_juego'
require_relative 'gestor_estados'
require_relative 'jugador'
require_relative 'mazo_sorpresas'
require_relative 'tablero'
require_relative 'sorpresa'
require_relative 'casilla_calle'
require_relative 'casilla_sorpresa'
require_relative 'casilla_impuesto'
require_relative 'casilla_juez'
require_relative 'casilla'
require_relative 'sorpresa_salir_carcel'
require_relative 'sorpresa_por_jugador'
require_relative 'sorpresa_por_casa_hotel'
require_relative 'sorpresa_pagar_cobrar'
require_relative 'sorpresa_ir_casilla'
require_relative 'sorpresa_ir_carcel'
require_relative 'sorpresa_especulador'
module Civitas
  class CivitasJuego
    
    #Atributo de clase - creado para evitar números mágicos en la creación del tablero
    @@casillacarcel=5
    
    #Constructures
    def initialize(nombres)
      @jugadores = Array.new
      for nombre in nombres
        @jugadores.push(Jugador.new(nombre))
      end
      
      @gestorestados = Gestor_estados.new()
      @estado = @gestorestados.estado_inicial
      @indicejugadoractual = Dado.instance.quien_empieza(@jugadores.size)
      @mazo = MazoSorpresas.new(false)
      
      inicializar_tablero(@mazo)
      inicializar_mazo_sorpresas(@tablero)
    end
    
    
    def inicializar_tablero(mazo)
      @tablero = Tablero.new(@@casillacarcel)
      
      #Ya está agregada la Salida en la posición 0 //nom, alquilerb, factorevaloriz, hipotebase, preccompra, precioedificar
      @tablero.aniade_casilla(CasillaCalle.new(TituloPropiedad.new("Calle Gonzalo Gallas", 75, 1.05, 400, 500, 350 )))
      @tablero.aniade_casilla(CasillaCalle.new(TituloPropiedad.new("Calle Severo Ochoa", 90, 1.1, 448, 560, 450 )))
      @tablero.aniade_casilla(CasillaSorpresa.new(mazo, "Sorpresa 1"))
      @tablero.aniade_casilla(CasillaCalle.new(TituloPropiedad.new("Calle Elvira", 105, 1.1, 504, 630, 550 )))
      #Aquí se añade la Cárcel      
      @tablero.aniade_casilla(CasillaCalle.new(TituloPropiedad.new("Calle Méndez Nuñez", 120, 1.2, 568, 710, 650)))
      @tablero.aniade_casilla(CasillaSorpresa.new(mazo, "Sorpresa 2"))
      @tablero.aniade_casilla(CasillaCalle.new(TituloPropiedad.new("Camino de Ronda", 135, 1.3, 640, 800, 750)))
      @tablero.aniade_casilla(CasillaCalle.new(TituloPropiedad.new("Calle Acera del Darro", 150, 1.3, 720, 900, 850)))
      @tablero.aniade_casilla(Casilla.new("Parking"))
      @tablero.aniade_casilla(CasillaCalle.new(TituloPropiedad.new("Gran Vía de Colón", 165, 1.4, 808, 1010, 950)))
      @tablero.aniade_casilla(CasillaImpuesto.new(250,"Impuesto"))
      @tablero.aniade_casilla(CasillaCalle.new(TituloPropiedad.new("Calle Ángel Ganivet", 180, 1.5, 904, 1130, 1050)))
      @tablero.aniade_casilla(CasillaCalle.new(TituloPropiedad.new("Calle Reyes Católicos", 195, 1.6, 1008, 1260, 1150)))
      @tablero.aniade_juez()
      @tablero.aniade_casilla(CasillaCalle.new(TituloPropiedad.new("Avenida de la Constitución", 210, 1.7, 1120, 1400, 1250)))
      @tablero.aniade_casilla(CasillaSorpresa.new(mazo, "Sorpresa 3"))
      @tablero.aniade_casilla(CasillaCalle.new(TituloPropiedad.new("Calle Recogidas", 225, 1.8, 1240, 1550, 1350 )))
      @tablero.aniade_casilla(CasillaCalle.new(TituloPropiedad.new("Paseo de los Tristes", 250, 2.0, 1368, 1710, 1450 )))
    end
    
    
    def inicializar_mazo_sorpresas(tablero)
     #@mazo.al_mazo(SorpresaSalirCarcel.new(@mazo))
    # @mazo.al_mazo(SorpresaIrCasilla.new(tablero, 12, "Debes pagar tus impuestos. Dirígete a la casilla 12."))
     @mazo.al_mazo(SorpresaIrCarcel.new(tablero))
    # @mazo.al_mazo(SorpresaIrCasilla.new(tablero, @@casillacarcel,"Dirígete a la carcel de visita"))
    # @mazo.al_mazo(SorpresaIrCasilla.new( tablero, 18, "Están preguntando por ti en las oficinas de Recogidas. Dirigete a la casilla 18"))
    # @mazo.al_mazo(SorpresaPagarCobrar.new( 500, "Has ganado la lotería. Recibes 500"))
    # @mazo.al_mazo(SorpresaPagarCobrar.new( -500, "Debes devolver 500 a hacienda."))
    # @mazo.al_mazo(SorpresaPorCasaHotel.new( 25, "Tus bienes se han revalorizado. Recibes 25 por cada construcción."))
   #  @mazo.al_mazo(SorpresaPorCasaHotel.new( -25, "Debes pagar 25 por cada construcción debido a reparaciones y obras"))
   #  @mazo.al_mazo(SorpresaPorJugador.new( 100, "Es tu cumpleaños, recibes 100 de cada jugador"))
   #  @mazo.al_mazo(SorpresaPorJugador.new( -60, "Tienes deudas pendientes. Das 60 a cada jugador"))
     @mazo.al_mazo(SorpresaEspeculador.new( 500 ))
    end
    
    
    def contabilizar_pasos_por_salida(jugadoractual)
      while (@tablero.porsalida > 0 )
        jugadoractual.pasa_por_salida
      end
    end
    
    
    def pasar_turno
      @indicejugadoractual = (@indicejugadoractual + 1) % @jugadores.size
    end
    
    
    def siguiente_paso_completado(operacion)
      @estado = @gestorestados.siguiente_estado(get_jugador_actual, @estado, operacion)
      
    end
    
    
    def cancelar_hipoteca(ip)
      get_jugador_actual.cancelar_hipoteca(ip)
    end
    
    
    def construir_casa(ip)
      get_jugador_actual.construir_casa(ip)
    end
    
    
    def construir_hotel(ip)
      get_jugador_actual.construir_hotel(ip)
    end
    
    
    def final_del_juego
      comprobacion = false
      for jugador in @jugadores
        if (jugador.en_bancarrota())
          comprobacion = true
        end
      end
      comprobacion
    end
    
    
    def get_casilla_actual
      @tablero.get_casilla(get_jugador_actual.num_casilla_actual)
    end
    
    
    def get_jugador_actual
      @jugadores[@indicejugadoractual]
    end
    
    
    def hipotecar(ip)
      get_jugador_actual.hipotecar(ip)
    end
    
    
    def infojugadortexto
      return get_jugador_actual.to_s
    end
    
    
    def ranking
      @jugadores.sort.reverse
    end
    
    
    def salir_carcel_pagando()
      get_jugador_actual.salir_carcel_pagando
    end
    
    
    def salir_carcel_tirando()
      get_jugador_actual.salir_carcel_tirando
    end
    
    
    def vender(ip)
      get_jugador_actual.vender(ip)
    end
    
    
    def comprar
      jugadoractual = get_jugador_actual
      casilla = get_casilla_actual##
      titulo = casilla.titulopropiedad()
      res = jugadoractual.comprar(titulo)
    end
    
    
    def siguiente_paso()
      jugadoractual = get_jugador_actual()
      operacion = @gestorestados.operaciones_permitidas(jugadoractual, @estado)
      if(operacion == Operaciones_juego::PASAR_TURNO)
        pasar_turno()
        siguiente_paso_completado(operacion)
      elsif(operacion == Operaciones_juego::AVANZAR)
        avanza_jugador()
        siguiente_paso_completado(operacion)
      end
      operacion
    end
    
    
    def avanza_jugador
      jugadoractual = get_jugador_actual()
      posicionactual = jugadoractual.num_casilla_actual()
      tirada = Dado.instance.tirar()
      posicionnueva = @tablero.nueva_posicion(posicionactual, tirada)
      casilla = @tablero.get_casilla(posicionnueva)
      contabilizar_pasos_por_salida(jugadoractual)
      jugadoractual.mover_a_casilla(posicionnueva)
      casilla.recibe_jugador(@indicejugadoractual, @jugadores)
      contabilizar_pasos_por_salida(jugadoractual)
    end
    
  end
end

File: ProyectoCivitasJava/src/civitas/CivitasJuego.java
/**
 * Proyecto Civitas
 * Sergio Hervás Cobo
 * Javier Suárez Lloréns
 * Programación y Diseño Orientado a Objetos
 * 2º B3 - Grado en Ingeniería Informática
 * E.T.S.I. Informática y de Telecomunicación
 * Universidad de Granada
 */
package civitas;

import GUI.Dado;
import java.util.ArrayList;
import java.util.Collections;

public class CivitasJuego {
    
    private final static int POSICION_CARCEL = 5;
    //Atributos de instancia
    private int indiceJugadorActual;
    
    //Atributos de referencia
    private Tablero tablero;
    private MazoSorpresas mazo;
    private EstadosJuego estado;
    private ArrayList <Jugador> jugadores;
    private GestorEstados gestorEstados;
    
    //Métodos       
    public CivitasJuego(ArrayList <String> nombres){
        //Inicializamos el ArrayList jugadores y 
        //asignamos los jugadores de la partida y sus nombres
        jugadores = new ArrayList<>(); 
        for (int k = 0; k < nombres.size(); k++){
          jugadores.add(new Jugador(nombres.get(k)));
        }
        
        gestorEstados = new GestorEstados();
        estado = gestorEstados.estadoInicial();
        
        indiceJugadorActual = Dado.getInstance().quienEmpieza(jugadores.size());
        
        mazo = new MazoSorpresas(false);

        inicializarTablero(mazo);
        inicializarMazoSorpresas(tablero);
   
    }
    
    private void avanzaJugador(){
        Jugador jugadorActual=jugadores.get(indiceJugadorActual);
        int posicionActual = jugadorActual.getNumCasillaActual();
        int tirada = Dado.getInstance().tirar();
        int posicionNueva = tablero.nuevaPosicion(posicionActual, tirada);
        Casilla casilla = tablero.getCasilla(posicionNueva);
        contabilizarPasosPorSalida(jugadores.get(indiceJugadorActual));
        jugadorActual.moverACasilla(posicionNueva);
        casilla.recibeJugador(indiceJugadorActual, jugadores);
        contabilizarPasosPorSalida(jugadores.get(indiceJugadorActual));
    }
    
    public boolean cancelarHipoteca(int ip){
        return jugadores.get(indiceJugadorActual).cancelarHipoteca(ip);
    }
    
    public boolean comprar(){

        Jugador jugadorActual = jugadores.get(indiceJugadorActual);
        int numCasillaActual = jugadorActual.getNumCasillaActual();
        CasillaCalle casilla = (CasillaCalle) tablero.getCasilla(numCasillaActual); //Casting
        TituloPropiedad titulo = casilla.getTituloPropiedad();
        boolean res = jugadorActual.comprar(titulo);
        
        return res;
        
    }
    
    public boolean construirCasa(int ip){
        return jugadores.get(indiceJugadorActual).construirCasa(ip);   
    }
    
    public boolean construirHotel(int ip){
        return jugadores.get(indiceJugadorActual).construirHotel(ip); 
    }
    
    private void contabilizarPasosPorSalida(Jugador jugadorActual){
        while(tablero.getPorSalida()>0){
            jugadorActual.pasaPorSalida();
        }
    }
    
    public boolean finalDelJuego(){
        boolean comprobacion = false;
        for(int k = 0; k < jugadores.size(); k++){
            if(jugadores.get(k).enBancarrota())
                comprobacion = true;
        }
        return comprobacion;
    }
    
    public Casilla getCasillaActual(){
        return tablero.getCasilla(jugadores.get(indiceJugadorActual).getNumCasillaActual());
        
    }
    
    public Jugador getJugadorActual(){
        return jugadores.get(indiceJugadorActual);  
    }
    
    public boolean hipotecar(int ip){
       return jugadores.get(indiceJugadorActual).hipotecar(ip);
    }
    
    public String infoJugadorTexto(){
        return jugadores.get(indiceJugadorActual).toString();
        
    }
    
    private void inicializarMazoSorpresas(Tablero tablero){
        mazo.alMazo(new SorpresaIrCarcel(tablero));
        mazo.alMazo(new SorpresaIrCasilla(tablero, POSICION_CARCEL, "Dirígete a la carcel de visita"));
        mazo.alMazo(new SorpresaIrCasilla(tablero, 12, "Debes pagar tus impuestos. Dirígete a la casilla 12."));
        mazo.alMazo(new SorpresaIrCasilla(tablero, 18, "Están preguntando por ti en las oficinas de Recogidas. Dirigete a la casilla 18"));
        mazo.alMazo(new SorpresaPagarCobrar( 500, "Has ganado la lotería. Recibes 500"));
        mazo.alMazo(new SorpresaPagarCobrar( -500, "Debes devolver 500 a hacienda."));
        mazo.alMazo(new SorpresaPorCasaHotel( 25, "Tus bienes se han revalorizado. Recibes 25 por cada construcción."));
        mazo.alMazo(new SorpresaPorCasaHotel(-25, "Debes pagar 25 por cada construcción debido a reparaciones y obras"));
        mazo.alMazo(new SorpresaPorJugador( 100, "Es tu cumpleaños, recibes 100 de cada jugador"));
        mazo.alMazo(new SorpresaPorJugador( -60, "Tienes deudas pendientes. Das 60 a cada jugador"));
        mazo.alMazo(new SorpresaSalirCarcel( mazo));
        mazo.alMazo(new SorpresaEspeculador( 500));
    
    }
    
    private void inicializarTablero(MazoSorpresas mazo){
        tablero = new Tablero(POSICION_CARCEL);
        //Ya está agregada la Salida en la posición 0
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Gonzalo Gallas",          75,  1.05f, 400 , 500 , 350  )));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Severo Ochoa",            90,  1.1f,  448 , 560 , 450  )));
        tablero.añadeCasilla(new CasillaSorpresa(mazo, "Sorpresa 1"));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Elvira",                  105, 1.1f,  504 , 630 , 550  )));
        //Aquí se añade la Cárcel
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Méndez Nuñez",            120, 1.2f,  568 , 710 , 650  )));
        tablero.añadeCasilla(new CasillaSorpresa(mazo, "Sorpresa 2"));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Camino de Ronda",               135, 1.3f,  640 , 800 , 750  )));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Acera del Darro",         150, 1.3f,  720 , 900 , 850  )));
        tablero.añadeCasilla(new Casilla("Parking"));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Gran Vía de Colón",             165, 1.4f,  808 , 1010, 950  )));
        tablero.añadeCasilla(new CasillaImpuesto(250f,"Impuesto de luz"));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Ángel Ganivet",           180, 1.5f,  904 , 1130, 1050 )));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Reyes Católicos",         195, 1.6f,  1008, 1260, 1150 )));
        tablero.añadeJuez();
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Avenida de la Constitución",    210, 1.7f,  1120, 1400, 1250 )));
        tablero.añadeCasilla(new CasillaSorpresa(mazo, "Sorpresa 3"));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Calle Recogidas",               225, 1.8f,  1240, 1550, 1350 )));
        tablero.añadeCasilla(new CasillaCalle(new TituloPropiedad("Paseo de los Tristes",          240, 2.0f,  1368, 1710, 1450 )));
        
    }
    
    private void pasarTurno(){
        indiceJugadorActual = (indiceJugadorActual+1)%jugadores.size();
        
    }
    
    
    public ArrayList <Jugador> ranking(){
        
        //Creamos un nuevo ArrayList de jugadores para ordenarlos
        ArrayList <Jugador> ranking;
        ranking = new ArrayList<>();
        for(int i = 0; i < jugadores.size();i++){
            ranking.add(jugadores.get(i));
        }
        
        //Llamamos al método de comparar de la clase Jugador mediante Collections
        Collections.sort(ranking, Collections.reverseOrder());
        return ranking;
    }
    
    public boolean salirCarcelPagando(){
        return jugadores.get(indiceJugadorActual).salirCarcelPagando();
        
    }
    
    public boolean salirCarcelTirando(){
        return jugadores.get(indiceJugadorActual).salirCarcelTirando();
    }
    
    public OperacionesJuego siguientePaso(){
        Jugador jugadorActual = jugadores.get(indiceJugadorActual);
        OperacionesJuego operacion = gestorEstados.operacionesPermitidas(jugadorActual, estado);
        
        if(operacion == OperacionesJuego.PASAR_TURNO){
            pasarTurno();
            siguientePasoCompletado(operacion);
        }
        else if (operacion == OperacionesJuego.AVANZAR){
            avanzaJugador();
            siguientePasoCompletado(operacion);
        }
        return operacion;
    }
    
    public void siguientePasoCompletado(OperacionesJuego operacion){
        estado = gestorEstados.siguienteEstado(jugadores.get(indiceJugadorActual), estado, operacion);
        
    }
    
    public boolean vender (int ip){
        return jugadores.get(indiceJugadorActual).vender(ip);
        
    }
    
    
    
}
